file-generating/URL-to-PDF api/README.md

```diff
@@ -0,0 +1,166 @@
+
+Singularity final.sif:/home/ysub> ps auxf
+USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
+root       461  0.0  0.0  18204  3188 pts/1    S    17:58   0:00 /bin/bash --norc
+root       468  0.0  0.0  36640  2880 pts/1    R+   17:59   0:00  \_ ps auxf
+root         1  0.0  0.1 565392 12144 ?        Sl   15:10   0:00 sinit
+root        16  0.0  0.4 1113904 39492 ?       Sl   15:10   0:00 npm
+root        26  0.0  0.0   4296   752 ?        S    15:10   0:00  \_ sh -c nodemon --watch ./src -e js src/index.js
+root        27  0.0  0.5 1179476 40312 ?       Sl   15:10   0:00      \_ node /pdf_server/node_modules/.bin/nodemon --watch ./src -e js src/index.js
+root        39  0.0  0.7 936444 61220 ?        Sl   15:10   0:02          \_ /usr/local/bin/node src/index.js
+
+Singularity final.sif:/home/ysub> exit
+```
+### Enhancing the Definition File
+
+Now that we have confirmation that the server is working, let’s make it a little cleaner. It’s difficult to remember the exact `curl` command and URL syntax each time you want to request a PDF, so this section will demonstrate enhancing of definition file to automate it. To do that, we can use Standard Container Integration Format (SCIF) apps, that are integrated directly into singularity. If you haven’t already, check out the [Scientific Filesystem documentation](https://sci-f.github.io/) to come up to speed.
+
+First off, we’re going to move the installation of the url-to-pdf into an app, so that there is a designated spot to place output files. To do that, we want to add a section to our definition file to build the server:
+```
+%appinstall pdf_server
+    git clone https://github.com/alvarcarto/url-to-pdf-api.git pdf_server
+    cd pdf_server
+    npm install
+    chmod -R 0755 .
+```
+And update our `startscript` to point to the app location:
+
+```
+%startscript
+    cd "${APPROOT_pdf_server}/pdf_server"
+    # Use nohup and /dev/null to completely detach server process from terminal
+    nohup npm start > /dev/null 2>&1 < /dev/null &
+```
+
+Now we want to define the pdf_client app, which we will run to send the requests to the server:
+```
+%apprun pdf_client
+    if [ -z "${1:-}" ]; then
+        echo "Usage: singularity run --app pdf <instance://name> <URL> [output file]"
+        exit 1
+
+    fi
+    curl -o "${SINGULARITY_APPDATA}/output/${2:-output.pdf}" "${URL}:${PORT}/api/render?url=${1}"
+```
+
+As you can see, the `pdf_client` app checks to make sure that the user provides at least one argument. Find the full def file [here](url-to-pdf-edited.def).
+
+Create the container as before. The `--force` option will overwrite the old container:
+```
+$ sudo singularity build --force url-to-pdf.sif url-to-pdf-edited.def
+```
+Now that we have an output directory in the container, we need to expose it to the host using a bind mount. Once we’ve rebuilt the container, make a new directory called `out` for the generated PDFs to go.
+```
+$ mkdir out
```

Comment:
Need to add where should the user be located to create this directory

Reply:
Can I write - "From your home directory"?

Reply:
sounds about right, thanks! @sushma-98